module/data/item/templates/activities.mjs

```diff
@@ -32,7 +33,55 @@ export default class ActivitiesTemplate extends SystemDataModel {
    * @param {object} source  Candidate source data to migrate.
    */
   static migrateActivities(source) {
-    ActivitiesTemplate.#migrateUses(source);
+    if ( source.system ) ActivitiesTemplate.#migrateUses(source.system);
+    // TODO: Handle migration of data kept on spells and inferred by activities
+    // TODO: Handle migration of base damage, range, & ammunition on weapons
+    if ( this.#shouldMigrateActivity(source) ) this.#migrateInitialActivity(source);
+  }
+
+  /* -------------------------------------------- */
+
+  /**
+   * Method to determine whether the activity creation migration should be performed. This migration should only be
+   * performed on whole item data rather than partial updates, so check to ensure all of the necessary data is present.
+   * @param {object} source  Candidate source data to migrate.
+   * @returns {boolean}
+   */
+  static #shouldMigrateActivity(source) {
+    // If source data doesn't have effects or system data, then is only a partial migration
+    if ( !source.effects || !source.system ) return false;
```

Comment:
Rather than relying on this heuristic, couldn't we instead invoke the activity migration from the `migrations.mjs` `migrateItemData`? Then we can be sure that it's only being run on full source data.

Also, should we add a check against `_stats.systemVersion`? I'm thinking about certain 'base' data like damage or range on weapons that might be accidentally re-migrated in some cases. Maybe it's impossible, but I don't think it can hurt to add this guard.

Reply:
If we run during `migrateItemData`, then all modules have to be explicitly migrated by their creators before they can be used by players. This ensures that those modules are usable on day 1.

We can add a `systemVersion` check to be doubly sure, but since we only perform the migration if `system.activities` is absent or empty, it should ensure that the migration doesn't occur twice.

Reply:
> If we run during `migrateItemData`, then all modules have to be explicitly migrated by their creators before they can be used by players. This ensures that those modules are usable on day 1.

That's always been the case though for any concrete migrations. I think there are a couple of things we'd need to watch out for with this in-memory migration:
1. It's not immune to false positives. It's possible (though unusual) to perform an update like `item.update({ "system.level": 9, effects: [{ /* ... */ }] })` which I think would result in spawning a new Activity because `source.activities` would be `undefined` for that update delta.
2. If you open an Item in a compendium that has not had `migrateCompendium` run on it, and then try to update it without importing (if it's in a personal shared compendium for example), then the delta sent to the server will not contain any of the migrated data that only exists in-memory at that time, it will contain only what changed. Upon refreshing, the previously-migrated activity data would be wiped, and you'd be left with only the fields that were updated.

We might be able to mitigate some of this by moving the migration to `_initializeSource`, which isn't called on update deltas, but I think (2) is always an issue.

> We can add a `systemVersion` check to be doubly sure, but since we only perform the migration if `system.activities` is absent or empty, it should ensure that the migration doesn't occur twice.

It's possible to create an Item with no activities on it though, no? Or delete any auto-generated ones?

Reply:
> It's not immune to false positives. It's possible (though unusual) to perform an update like `item.update({ "system.level": 9, effects: [{ /* ... */ }] })` which I think would result in spawning a new Activity because `source.activities` would be `undefined` for that update delta.

True, we could resolve that by adding more edge data points that would be guaranteed to be in the full data but not in a partial update like `ownership`.

> If you open an Item in a compendium that has not had `migrateCompendium` run on it, and then try to update it without importing (if it's in a personal shared compendium for example), then the delta sent to the server will not contain any of the migrated data that only exists in-memory at that time, it will contain only what changed. Upon refreshing, the previously-migrated activity data would be wiped, and you'd be left with only the fields that were updated.

Hmm, that is part of the reason I didn't include a `systemVersion` check to start with.

> It's possible to create an Item with no activities on it though, no? Or delete any auto-generated ones?

Yes, but a newly created item would be missing `actionType` and `activation.type` and a world item migrated in that way would have them removed during `persistSourceMigration` (I believe). That would probably be an issue on a user's module compendium that never had `migrateCompendium` run on it if they try to delete the auto-generated activity.